Flowaggregator panic: runtime error

**Describe the bug**
flow-aggregator pod crashes fairly infrequently.

```yaml
NAMESPACE            NAME                                            READY   STATUS    RESTARTS   AG
flow-aggregator      flow-aggregator-69f476cf7c-mbcgv                1/1     Running   22         9d
```
`kubectl describe` does not give much information to go on. 

```yaml
Last State:     Terminated
      Reason:       Error
      Exit Code:    2
      Started:      Tue, 12 Jan 2021 13:52:00 +0000
      Finished:     Tue, 12 Jan 2021 14:08:25 +0000
```

Using kubectl logs with previous flag I can grab the logs but often not all of the traceback. 

`kubectl logs flow-aggregator-ddd46cfc7-h4kff -n flow-aggregator -p`

I have included the few lines above the "panic" to show that it is not the same preceding issue causing it, so it seems. The runtime error sometimes has different output such as actual segmentation fault information in a few instances.

**Actual behavior**

`kubectl logs flow-aggregator-ddd46cfc7-h4kff -n flow-aggregator -p`
```
I0112 13:52:00.592685       1 flow-aggregator.go:27] Flow aggregator starting...
I0112 13:52:00.597287       1 tcp.go:38] Start tcp collecting process on [::]:4739
E0112 14:00:55.171757       1 tcp.go:90] collector only supports IPFIX (v10); invalid version 2923 received
panic: runtime error: slice bounds out of range [:232] with capacity 95

goroutine 53 [running]:
github.com/vmware/go-ipfix/pkg/collector.(*CollectingProcess).handleTCPClient.func1(0x91b940, 0xc000107258, 0xc0002d2000, 0xc0019812c0, 0xf)
	/go/pkg/mod/github.com/vmware/go-ipfix@v0.4.2/pkg/collector/tcp.go:88 +0x68a
created by github.com/vmware/go-ipfix/pkg/collector.(*CollectingProcess).handleTCPClient
	/go/pkg/mod/github.com/vmware/go-ipfix@v0.4.2/pkg/collector/tcp.go:65 +0x18a
```
different crash below
```
W0112 14:28:55.161455       1 aggregate.go:247] This field with name destinationClusterIPv4 should not have been filled with value <nil> in existing record.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x6e38ad]

goroutine 21 [running]:
github.com/vmware/go-ipfix/pkg/intermediate.(*AggregationProcess).aggregateRecords(0xc0000a2150, 0x91b820, 0xc002a33230, 0x91b820, 0xc000216870, 0x100, 0x88fe4a, 0x12)
	/go/pkg/mod/github.com/vmware/go-ipfix@v0.4.2/pkg/intermediate/aggregate.go:312 +0x5ad
github.com/vmware/go-ipfix/pkg/intermediate.(*AggregationProcess).addOrUpdateRecordInMap(0xc0000a2150, 0xc000e07d70, 0x91b820, 0xc002a33230, 0x0, 0x0)
	/go/pkg/mod/github.com/vmware/go-ipfix@v0.4.2/pkg/intermediate/aggregate.go:176 +0x2db
github.com/vmware/go-ipfix/pkg/intermediate.(*AggregationProcess).AggregateMsgByFlowKey(0xc0000a2150, 0xc001022dc0, 0x2, 0x2)
	/go/pkg/mod/github.com/vmware/go-ipfix@v0.4.2/pkg/intermediate/aggregate.go:118 +0x130
github.com/vmware/go-ipfix/pkg/intermediate.(*worker).start.func1(0xc0000f3680)
	/go/pkg/mod/github.com/vmware/go-ipfix@v0.4.2/pkg/intermediate/worker.go:35 +0x111
created by github.com/vmware/go-ipfix/pkg/intermediate.(*worker).start
	/go/pkg/mod/github.com/vmware/go-ipfix@v0.4.2/pkg/intermediate/worker.go:26 +0x3f
```
different crash below
```
E0113 04:30:57.224860       1 process.go:195] Error when closing connection to collector: close udp 10.20.0.248:51440->10.103.98.74:4739: use of closed network connection
panic: runtime error: slice bounds out of range [:249] with capacity 65

goroutine 46 [running]:
github.com/vmware/go-ipfix/pkg/collector.(*CollectingProcess).handleTCPClient.func1(0x91b940, 0xc0171fe000, 0xc000266000, 0xc019a7a010, 0xe)
	/go/pkg/mod/github.com/vmware/go-ipfix@v0.4.2/pkg/collector/tcp.go:88 +0x68a
created by github.com/vmware/go-ipfix/pkg/collector.(*CollectingProcess).handleTCPClient
	/go/pkg/mod/github.com/vmware/go-ipfix@v0.4.2/pkg/collector/tcp.go:65 +0x18a
```

**Versions:**

    Antrea version 0.12
    Kubernetes version:
```
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.1", GitCommit:"c4d752765b3bbac2237bf87cf0b1c2e307844666", GitTreeState:"clean", BuildDate:"2020-12-18T12:09:25Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.4", GitCommit:"d360454c9bcd1634cf4cc52d1867af5491dc9c5f", GitTreeState:"clean", BuildDate:"2020-11-11T13:09:17Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
```
    Container runtime: Docker
    Linux kernel version: 5.4.0-47-generic
    Openvswitch


**To Reproduce**

Run container with 
```
AntreaProxy: true
Traceflow: true
FlowExporter: true
flowCollectorAddr: "flow-aggregator.flow-aggregator.svc:4739:tcp"
```

Have flow agg config setup to forward to ELK logstash pod on udp. 
```yaml
apiVersion: v1
data:
  flow-aggregator.conf: |
    externalFlowCollectorAddr: "logstash.elk-flow-collector.svc.cluster.local:4739:udp"
    flowExportInterval: 60s
```

**Expected**
No restarts / terminations.

**Additional context**
I understand a packet capture may be of assistance here however I am unsure what the best content / method to share it would be.

--- a/ci/kind/test-e2e-kind.sh
+++ b/ci/kind/test-e2e-kind.sh
@@ -92,7 +92,7 @@ if $np; then
     manifest_args="$manifest_args --np --tun vxlan"
 fi
 
-COMMON_IMAGES_LIST=("gcr.io/kubernetes-e2e-test-images/agnhost:2.8" "projects.registry.vmware.com/library/busybox" "projects.registry.vmware.com/antrea/nginx" "projects.registry.vmware.com/antrea/perftool" "projects.registry.vmware.com/antrea/ipfix-collector:v0.4.2")
+COMMON_IMAGES_LIST=("gcr.io/kubernetes-e2e-test-images/agnhost:2.8" "projects.registry.vmware.com/library/busybox" "projects.registry.vmware.com/antrea/nginx" "projects.registry.vmware.com/antrea/perftool" "projects.registry.vmware.com/antrea/ipfix-collector:v0.4.3")
 for image in "${COMMON_IMAGES_LIST[@]}"; do
     docker pull $image
 done

--- a/go.mod
+++ b/go.mod
@@ -42,7 +42,7 @@ require (
 	github.com/stretchr/testify v1.6.1
 	github.com/ti-mo/conntrack v0.3.0
 	github.com/vishvananda/netlink v1.1.0
-	github.com/vmware/go-ipfix v0.4.2
+	github.com/vmware/go-ipfix v0.4.3
 	golang.org/x/crypto v0.0.0-20200820211705-5c72a883971a
 	golang.org/x/exp v0.0.0-20190312203227-4b39c73a6495
 	golang.org/x/mod v0.4.0

--- a/pkg/agent/flowexporter/exporter/exporter.go
+++ b/pkg/agent/flowexporter/exporter/exporter.go
@@ -190,32 +190,24 @@ func (exp *flowExporter) initFlowExporter(collectorAddr string, collectorProto s
 		}
 	}
 
-	// TODO: This code can be further simplified by changing the go-ipfix API to accept
-	// collectorAddr and collectorProto instead of net.Addr input.
 	var expInput exporter.ExporterInput
 	if collectorProto == "tcp" {
-		collector, err := net.ResolveTCPAddr("tcp", collectorAddr)
-		if err != nil {
-			return err
-		}
 		// TCP transport does not need any tempRefTimeout, so sending 0.
 		// tempRefTimeout is the template refresh timeout, which specifies how often
 		// the exporting process should send the template again.
 		expInput = exporter.ExporterInput{
-			CollectorAddr:       collector,
+			CollectorAddress:    collectorAddr,
+			CollectorProtocol:   collectorProto,
 			ObservationDomainID: obsID,
 			TempRefTimeout:      0,
 			PathMTU:             0,
 			IsEncrypted:         false,
 		}
 	} else {
-		collector, err := net.ResolveUDPAddr("udp", collectorAddr)
-		if err != nil {
-			return err
-		}
 		// For UDP transport, hardcoding tempRefTimeout value as 1800s.
 		expInput = exporter.ExporterInput{
-			CollectorAddr:       collector,
+			CollectorAddress:    collectorAddr,
+			CollectorProtocol:   collectorProto,
 			ObservationDomainID: obsID,
 			TempRefTimeout:      1800,
 			PathMTU:             0,
@@ -452,7 +444,11 @@ func (exp *flowExporter) addRecordToSet(dataSet ipfix.IPFIXSet, record flowexpor
 				ie.Value = net.ParseIP("::")
 			}
 		case "destinationServicePort":
-			ie.Value = record.Conn.TupleOrig.DestinationPort
+			if record.Conn.DestinationServicePortName != "" {
+				ie.Value = record.Conn.TupleOrig.DestinationPort
+			} else {
+				ie.Value = uint16(0)
+			}
 		case "destinationServicePortName":
 			if record.Conn.DestinationServicePortName != "" {
 				ie.Value = record.Conn.DestinationServicePortName

--- a/pkg/flowaggregator/flowaggregator.go
+++ b/pkg/flowaggregator/flowaggregator.go
@@ -17,7 +17,6 @@ package flowaggregator
 import (
 	"fmt"
 	"hash/fnv"
-	"net"
 	"time"
 
 	"github.com/vmware/go-ipfix/pkg/collector"
@@ -129,6 +128,9 @@ var (
 
 const (
 	aggregationWorkerNum = 2
+	udpTransport         = "udp"
+	tcpTransport         = "tcp"
+	collectorAddress     = "0.0.0.0:4739"
 )
 
 type AggregatorTransportProtocol string
@@ -176,21 +178,20 @@ func genObservationID() (uint32, error) {
 }
 
 func (fa *flowAggregator) InitCollectingProcess() error {
-	var collectAddr net.Addr
 	var err error
 	var cpInput collector.CollectorInput
 	if fa.aggregatorTransportProtocol == AggregatorTransportProtocolTCP {
-		collectAddr, _ = net.ResolveTCPAddr("tcp", "0.0.0.0:4739")
 		cpInput = collector.CollectorInput{
-			Address:       collectAddr,
+			Address:       collectorAddress,
+			Protocol:      tcpTransport,
 			MaxBufferSize: 65535,
 			TemplateTTL:   0,
 			IsEncrypted:   false,
 		}
 	} else {
-		collectAddr, _ = net.ResolveUDPAddr("udp", "0.0.0.0:4739")
 		cpInput = collector.CollectorInput{
-			Address:       collectAddr,
+			Address:       collectorAddress,
+			Protocol:      udpTransport,
 			MaxBufferSize: 1024,
 			TemplateTTL:   0,
 			IsEncrypted:   false,
@@ -221,26 +222,20 @@ func (fa *flowAggregator) initExportingProcess() error {
 	// externalFlowCollectorAddr and externalFlowCollectorProto instead of net.Addr input.
 	var expInput exporter.ExporterInput
 	if fa.externalFlowCollectorProto == "tcp" {
-		collector, err := net.ResolveTCPAddr("tcp", fa.externalFlowCollectorAddr)
-		if err != nil {
-			return err
-		}
 		// TCP transport does not need any tempRefTimeout, so sending 0.
 		expInput = exporter.ExporterInput{
-			CollectorAddr:       collector,
+			CollectorAddress:    fa.externalFlowCollectorAddr,
+			CollectorProtocol:   fa.externalFlowCollectorProto,
 			ObservationDomainID: obsID,
 			TempRefTimeout:      0,
 			PathMTU:             0,
 			IsEncrypted:         false,
 		}
 	} else {
-		collector, err := net.ResolveUDPAddr("udp", fa.externalFlowCollectorAddr)
-		if err != nil {
-			return err
-		}
 		// For UDP transport, hardcoding tempRefTimeout value as 1800s. So we will send out template every 30 minutes.
 		expInput = exporter.ExporterInput{
-			CollectorAddr:       collector,
+			CollectorAddress:    fa.externalFlowCollectorAddr,
+			CollectorProtocol:   fa.externalFlowCollectorProto,
 			ObservationDomainID: obsID,
 			TempRefTimeout:      1800,
 			PathMTU:             0,

--- a/test/e2e/framework.go
+++ b/test/e2e/framework.go
@@ -93,7 +93,7 @@ const (
 	busyboxImage        = "projects.registry.vmware.com/library/busybox"
 	nginxImage          = "projects.registry.vmware.com/antrea/nginx"
 	perftoolImage       = "projects.registry.vmware.com/antrea/perftool"
-	ipfixCollectorImage = "projects.registry.vmware.com/antrea/ipfix-collector:v0.4.2"
+	ipfixCollectorImage = "projects.registry.vmware.com/antrea/ipfix-collector:v0.4.3"
 	ipfixCollectorPort  = "4739"
 
 	nginxLBService = "nginx-loadbalancer"

